### PR TITLE
Continue the implementation of the service registry with the custom ERC20 token security

### PR DESCRIPTION
**Pull request by kupermind**
- Added `update()` function in the new `ServiceManager` contract;
- Adding tests.

### CODE REVIEW

**Comment by 77ph** [comment]
Everything found is in internal audit.

**Review by 77ph — APPROVED** [review]

File: contracts/ServiceManagerToken.sol [hunk 1]
```diff
@@ -32,14 +32,22 @@ contract ServiceManagerToken is GenericManager {
     // Operator whitelist address
     address public operatorWhitelist;
 
+    /// @dev ServiceRegistryTokenUtility constructor.
+    /// @param _serviceRegistry Service Registry contract address.
+    /// @param _serviceRegistryTokenUtility Service Registry Token Utility contract address.
     constructor(address _serviceRegistry, address _serviceRegistryTokenUtility) {
```
Comment (mariapiamo) [inline contracts/ServiceManagerToken.sol:38]: Ok, here we will also add the operator whitelist address and the corresponding non-zero check

File: contracts/ServiceManagerToken.sol [hunk 2]
```diff
@@ -107,22 +114,56 @@ contract ServiceManagerToken is GenericManager {
...
+            revert ZeroAddress();
+        }
+
+        if (token == ETH_TOKEN_ADDRESS) {
+            // Call the original ServiceRegistry contract function
+            IService(serviceRegistry).update(msg.sender, configHash, agentIds, agentParams, threshold, serviceId);
+            // Reset the service token-based data
+            // This function still needs to be called as the previous token could be a custom ERC20 token
+            IServiceTokenUtility(serviceRegistryTokenUtility).resetServiceToken(serviceId);
+        } else {
```
Comment (mariapiamo) [inline contracts/ServiceManagerToken.sol:144]: The following will be fixed

File: contracts/ServiceRegistryTokenUtility.sol [hunk 3]
```diff
@@ -148,28 +155,35 @@ contract ServiceRegistryTokenUtility is IErrorsRegistries {
     }
 
     /// @dev Creates a record with the token-related information for the specified service.
+    /// @notice We assume that the token is checked for being a non-zero address and a non-ETH address representation
+    ///         outside of this function. Here we optimistically check for the token to have a specific `balanceOf`
```
Comment (mariapiamo) [inline contracts/ServiceRegistryTokenUtility.sol:159]: what do you mean by "we optimistically check"?

Comment (mariapiamo) [reply contracts/ServiceRegistryTokenUtility.sol:159]: All the notice is not super clear to me, maybe can be rephrased a bit

File: contracts/ServiceRegistryTokenUtility.sol [hunk 4]
```diff
@@ -148,28 +155,35 @@ contract ServiceRegistryTokenUtility is IErrorsRegistries {
...
-        if (token == address(0)) {
-            revert ZeroAddress();
-        }
         // TODO Check for the token ERC20 formality
         
         uint256 securityDeposit;
         // Service is newly created and all the array lengths are checked by the original ServiceRegistry create() function
         for (uint256 i = 0; i < agentIds.length; ++i) {
-            // Check for a non-zero bond value
+            // Check for a non-zero bond value and skip those with zeros (possible when updating a service)
```
Comment (mariapiamo) [inline contracts/ServiceRegistryTokenUtility.sol:184]: when this function is used to `update` the service, we can have some bonds equal to zero. But we still need to check that there is at least one i such that bond[i] is not zero, correct?

Comment (mariapiamo) [reply contracts/ServiceRegistryTokenUtility.sol:184]: Ah no, I see from the next PR that a check is made at the serviceManager level. Correct?

File: contracts/ServiceManagerToken.sol [hunk 5]
```diff
@@ -107,22 +114,56 @@ contract ServiceManagerToken is GenericManager {
...
     ) external returns (bool)
     {
-        return IService(serviceRegistry).update(msg.sender, configHash, agentIds, agentParams,
-            threshold, serviceId);
+        // Check for the zero address
+        if (token == address(0)) {
+            revert ZeroAddress();
+        }
+
+        if (token == ETH_TOKEN_ADDRESS) {
```
Comment (mariapiamo) [inline contracts/ServiceManagerToken.sol:138]: here we will add the zero-bond check

File: contracts/ServiceManagerToken.sol [hunk 6]
```diff
@@ -107,22 +114,56 @@ contract ServiceManagerToken is GenericManager {
...
+            // This function still needs to be called as the previous token could be a custom ERC20 token
+            IServiceTokenUtility(serviceRegistryTokenUtility).resetServiceToken(serviceId);
+        } else {
+            // Wrap agent params with just 1 WEI bond going to the original ServiceRegistry contract,
+            // and actual token bonds being recorded with the ServiceRegistryTokenUtility contract
+            uint256 numAgents = agentParams.length;
+            uint256[] memory bonds = new uint256[](numAgents);
+            for (uint256 i = 0; i < numAgents; ++i) {
+                // Copy actual bond values for each agent Id that has at least one slot in the updated service
+                if (agentParams[i].slots > 0) {
```
Comment (mariapiamo) [inline contracts/ServiceManagerToken.sol:151]: Now, if agentParams[i].slots > 0, we also need to make sure that agentParams[i].bonds> 0

File: contracts/ServiceRegistryTokenUtility.sol (hunk 3 above)
Comment (mariapiamo) [reply contracts/ServiceRegistryTokenUtility.sol:159]: clarified in the next pr

File: contracts/ServiceRegistryTokenUtility.sol (hunk 4 above)
Comment (mariapiamo) [reply contracts/ServiceRegistryTokenUtility.sol:184]: yes

File: contracts/ServiceManagerToken.sol (hunk 6 above)
Comment (mariapiamo) [reply contracts/ServiceManagerToken.sol:151]: added in the next PR

File: contracts/ServiceManagerToken.sol (hunk 5 above)
Comment (mariapiamo) [reply contracts/ServiceManagerToken.sol:138]: added in the next PR

**Pull request merged**